core/src/test/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandlerTest.java

```diff
@@ -0,0 +1,385 @@
+
+import reactor.bus.Event;
+import reactor.bus.EventBus;
+
+@ExtendWith(MockitoExtension.class)
+class StopStartDownscaleGetRecoveryCandidatesHandlerTest {
+
+    private static final String MOCK_INSTANCEID_PREFIX = "i-";
+
+    private static final String MOCK_FQDN_PREFIX = "fqdn-";
+
+    private static final String INSTANCE_GROUP_NAME = "compute";
+
+    private static final Long STACK_ID = 1L;
+
+    private static final Integer ALL_INSTANCES_IN_HG_COUNT = 20;
+
+    private static final Integer RUNNING_INSTANCES_COUNT = 16;
+
+    private static final Integer UNHEALTHY_RUNNING_INSTANCES_COUNT = 3;
+
+    private static final String RUNTIME = "7.2.15";
+
+    @Mock
+    private EventBus eventBus;
+
+    @Mock
+    private StackService stackService;
+
+    @Mock
+    private CloudPlatformConnectors cloudPlatformConnectors;
+
+    @Mock
+    private CloudContext cloudContext;
+
+    @Mock
+    private CloudStack cloudStack;
+
+    @Mock
+    private CloudCredential cloudCredential;
+
+    @Mock
+    private CloudConnector cloudConnector;
+
+    @Mock
+    private InstanceConnector instanceConnector;
+
+    @Mock
+    private ClusterApiConnectors clusterApiConnectors;
+
+    @Mock
+    private ClusterApi clusterApi;
+
+    @Mock
+    private ClusterHealthService clusterHealthService;
+
+    @Mock
+    private DetailedHostStatuses detailedHostStatuses;
+
+    @Mock
+    private CloudInstanceIdToInstanceMetaDataConverter cloudInstanceIdToInstanceMetaDataConverter;
+
+    @Mock
+    private InstanceMetaDataToCloudInstanceConverter instanceMetaDataToCloudInstanceConverter;
+
+    @Mock
+    private InstanceMetaDataService instanceMetaDataService;
+
+    @Mock
+    private CloudbreakFlowMessageService flowMessageService;
+
+    @Mock
+    private RuntimeVersionService runtimeVersionService;
+
+    @Captor
+    private ArgumentCaptor<Event> argumentCaptor;
+
+    @InjectMocks
+    private StopStartDownscaleGetRecoveryCandidatesHandler underTest;
+
+    @BeforeEach
+    void setup() {
+        Authenticator authenticator = mock(Authenticator.class);
+        AuthenticatedContext ac = mock(AuthenticatedContext.class);
+        CloudPlatformVariant cloudPlatformVariant = mock(CloudPlatformVariant.class);
+
+        lenient().doReturn(cloudPlatformVariant).when(cloudContext).getPlatformVariant();
+        lenient().doReturn(cloudConnector).when(cloudPlatformConnectors).get(any(CloudPlatformVariant.class));
+        lenient().doReturn(authenticator).when(cloudConnector).authentication();
+        lenient().doReturn(ac).when(authenticator).authenticate(any(CloudContext.class), any(CloudCredential.class));
+        lenient().doReturn(instanceConnector).when(cloudConnector).instances();
+        lenient().doReturn(Optional.of(RUNTIME)).when(runtimeVersionService).getRuntimeVersion(anyLong());
+    }
+
+    @Test
+    void testType() {
+        assertEquals(StopStartDownscaleGetRecoveryCandidatesRequest.class, underTest.type());
+    }
+
```

Comment:
Would be useful to have some negative tests.
e.g. node in maint_mode, node is DECOMMISSIONED, node is UNHEALTHY - should not show up as a recovery candidate. This could probably piggyback on existing tests by mocking a few more instances into UNHEALTHY state. The test verification likely won't change much.

Reply:
Introduced negative test cases